### PR TITLE
Post author selector: try downshift

### PR DESCRIPTION
**Pull request by adamsilverstein**
## Description
This PR adds a selector implementation built with [downshift](https://github.com/paypal/downshift), an MIT licensed  `downshift` from PayPal which aims to provide "Primitives to build simple, flexible, WAI-ARIA compliant React autocomplete/dropdown/select/combobox components"

Addresses https://github.com/WordPress/gutenberg/issues/7384

## How has this been tested?
* Tried with small number and large number of users, added with `wp user generate --role=editor --count=12000`
* Tested using only keyboard to interact.
* Tested with voiceover, checking results announcements

## Types of changes
* include downshift, replacing author select
* grab 100 users by default from REST API, 

## Questions / Todo
* no special handling for smaller numbers of users, should we add that? eg use regular select element when < 50 users?
* aria strings should be passed thru our i18n function

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation.

### CODE REVIEW

**Comment by adamsilverstein** [comment]
@afercia appreciate a quick review of this approach as well when you have a moment!

File: editor/components/post-author/index.js [hunk 1]
```diff
@@ -5,45 +5,107 @@ import { __ } from '@wordpress/i18n';
...
+import apiRequest from '@wordpress/api-request';
 
 /**
  * Internal dependencies
  */
 import PostAuthorCheck from './check';
 
+/**
+ * External dependencies
+ */
```
Comment (youknowriad) [inline editor/components/post-author/index.js]: In general the external dependencies are defined above all other deps, but this is not an official guideline :)

File: editor/components/post-author/index.js [hunk 2]
```diff
@@ -5,45 +5,107 @@ import { __ } from '@wordpress/i18n';
...
+import Downshift from 'downshift';
+import { debounce } from 'underscore';
```
Comment (youknowriad) [inline editor/components/post-author/index.js]: Should we use `lodash` instead?

File: editor/components/post-author/index.js [hunk 3]
```diff
@@ -5,45 +5,107 @@ import { __ } from '@wordpress/i18n';
...
+		onUpdateAuthor( Number( id ) );
+	}
+
+	suggestAuthors( query, args ) {
+		if ( ! args.isOpen ) {
+			return;
+		}
+		const payload = '?search=' + encodeURIComponent( query );
+		this.setState( { searching: true } );
+		apiRequest( { path: '/wp/v2/users' + payload } ).done( ( results ) => {
```
Comment (youknowriad) [inline editor/components/post-author/index.js]: Should we leverage the data module instead? We already have a `users` reducer their which I believe should be refactored to be an entity and build support to searching/pagination into the entities. related #6395

**Review by youknowriad — COMMENTED** [review]
Thanks for trying to move this forward. Might be good to get some a11y insights on Downshift.

**Comment by afercia** [comment]
@adamsilverstein good morning, thanks for the ping. Looking just at the a11y side, I'm not particularly happy with what `downshift` does. Their implementation is basically wrong, especially with regards to `aria-selected`. It's not supposed to work that way.

When an option in the list gets highlighted, it should also get `aria-selected="true"`:
https://www.w3.org/TR/wai-aria-practices/#combobox
> In a combobox with a listbox, grid, or tree popup, when a suggested value is visually indicated as the currently selected value, the option, gridcell, row, or treeitem containing that value has aria-selected set to true.

Instead, they switch it to `true` only when it's the actual selected value. This is wrong. A correct usage of `aria-selected` is required to make the combobox work consistently across browsers/screen readers combinations. For example, the only reason why the highlighted option is announced with Safari + VoiceOver is because `downshift` uses an `aria-live` region to announce the option. Without that, Safari + VoiceOver wouldn't announce anything. However, a combobox is not supposed to use an `aria-live` region at all. With other browsers/screen readers combinations, the combobox works natively and the options get announced _plus_ the change in the `aria-live` region gets announced. This makes the highlighted option be announced twice:

screenshot 14

There are other minor quirks, for example:
- the input content sometimes disappears when tabbing 
- the number of results gets announced multiple times (see screenshot above), seems it gets a first set of results and then a second set and so on (may depend on slow response maybe?)
- other aria attributes that could be improved. Overall, from an a11y perspective, the Gutenberg auto-completers work better.

Aside: the component crashes in IE 11 because it uses `includes()`:

screenshot 15

Info for the ones willing to test `downshift` on their [demo page](http://downshift.netlify.com/?selectedKind=Examples&selectedStory=basic&full=0&addons=1&stories=1&panelRight=0): it doesn't work with Safari + VoiveOver because it's within an iframe; you have to [access the iframe directly](http://downshift.netlify.com/iframe.html?selectedKind=Examples&selectedStory=basic).

**Comment by karmatosed** [comment]
Design wise lets have something like this:

author

The only ponder I have with this is the closed section for author. I added it as this could be in future multiple authors and also there is the very real possibility of long author names.

This design uses the same pattern as links so makes a lot of sense.

**Comment by karmatosed** [comment]
[ edit ] Re-uploading as something weird happened to my images - sorry for second post on this.

I have a suggested design, firstly lets have at the top as this is a common use. I would like to not have the input next to 'author' as we could have very probably cases of long author names. Let's have it as a  section, this also allows it to have multiples in future.

The design itself takes a lot of inspiration from links and should work.

author1
author2

**Comment by adamsilverstein** [comment]
> Looking just at the a11y side, I'm not particularly happy with what downshift does. Their implementation is basically wrong, especially with regards to aria-selected. It's not supposed to work that way.

@afercia :( Maybe I should double back on my autocompleter based implementation. Could you try this PR as well? https://github.com/WordPress/gutenberg/pull/7385

**Comment by adamsilverstein** [comment]
@karmatosed Thanks for the design, that is great and I like how it feels to give the field more room. question: when there is already a user selected and i click in the box, should the field clear/show the placeholder?

**Comment by afercia** [comment]
Just noting the design of the field is one more case where the placeholder is used as replacement for a label. We've warned several times this is [far from ideal for accessibility](https://core.trac.wordpress.org/ticket/40331) and goes [against the HTML spec](https://www.w3.org/TR/html52/sec-forms.html#the-placeholder-attribute).

**Comment by adamsilverstein** [comment]
> Just noting the design of the field is one more case where the placeholder is used as replacement for a label. We've warned several times this is far from ideal for accessibility and goes against the HTML spec.

Can we still have a label instead of a placeholder then, say something like `Type to select user:` maybe on a line above the selector so there is still room. This does seem to duplicate some information, is there a way to add a label that doesn't display anything but still gets read correctly?

**Comment by afercia** [comment]
We're using a lot of labels visually hidden with `screen-reader-text`. The real point is the placeholder shouldn't be used that way ;)

**Comment by adamsilverstein** [comment]
Got it, thanks for clarifying.

**Comment by karmatosed** [comment]
> when there is already a user selected and i click in the box, should the field clear/show the placeholder?

Yes it should clear.

**Comment by aduth** [comment]
Is it likely this will be resumed, or can it be closed?

**Comment by adamsilverstein** [comment]
> Is it likely this will be resumed, or can it be closed?

Closing this. I'll add any further work on https://github.com/WordPress/gutenberg/pull/7385

**Pull request closed without merging**